### PR TITLE
Allow multiple containers in worker pod in case for ErrPodCompleted

### LINKED ISSUE #710: Multiple containers in worker pod is not allowed under certain circumstances
## Description

Kubernetes work fails with following error if multiple containers are in the  worker pod and `worker` container in the pod has completed in short time.

```text
ERROR: Remote unit failed: Error creating pod: expected 1 container in pod but there were 2
```

- ✅ Succeeded: Single container in the pod, `worker` container invokes `sleep 5`
- ✅ Succeeded: Single container in the pod, `worker` container invokes `sleep 0`
- ✅ Succeeded: Multiple containers in the pod, `worker` container invokes `sleep 5`
- 🚫 Failed: Multiple containers in the pod, `worker` container invokes `sleep 0`

## Version

Using Receptor built by `make receptor` on upstream devel branch.

```bash
$ receptorctl --socket /tmp/receptor.sock version
Warning: receptorctl and receptor are different versions, they may not be compatible
receptorctl  1.3.0
receptor     v1.3.0+g8f8481c
```

## Receptor configuration

```yaml
---
- local-only:
- node:
    id: demo
- log-level:
    level: debug
- control-service:
    service: control
    filename: /tmp/receptor.sock
- work-kubernetes:
    worktype: demo
    authmethod: kubeconfig
    allowruntimepod: true
    allowruntimeauth: true
```

## Steps to reproduce

```bash
kubectl create ns receptor
receptor -c receptor.yml
```

### ✅ Succeeded: Single container in the pod, `worker` container invokes `sleep 5`

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: demo
  namespace: receptor
spec:
  containers:
    - name: worker
      image: busybox
      command: ["sh", "-c", "sleep 5; echo WORKER DONE"]
  restartPolicy: Never
```

```bash
$ receptorctl --socket /tmp/receptor.sock work submit --rm --follow --no-payload --param secret_kube_config=@/etc/rancher/k3s/k3s.yaml --param secret_kube_pod=@./pod-single-5.yml demo
WORKER DONE
(pzXNjb6g, released)
```

### ✅ Succeeded: Single container in the pod, `worker` container invokes `sleep 0`

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: demo
  namespace: receptor
spec:
  containers:
    - name: worker
      image: busybox
      command: ["sh", "-c", "sleep 0; echo WORKER DONE"]
  restartPolicy: Never
```

```bash
$ receptorctl --socket /tmp/receptor.sock work submit --rm --follow --no-payload --param secret_kube_config=@/etc/rancher/k3s/k3s.yaml --param secret_kube_pod=@./pod-single-0.yml demo
WORKER DONE
(4ZfiaQ6c, released)
```

### ✅ Succeeded: Multiple containers in the pod, `worker` container invokes `sleep 5`

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: demo
  namespace: receptor
spec:
  containers:
    - name: worker
      image: busybox
      command: ["sh", "-c", "sleep 5; echo WORKER DONE"]
    - name: sidecar
      image: busybox
      command: ["sh", "-c", "sleep 15; echo SIDECAR DONE"]
  restartPolicy: Never
```

```bash
$ receptorctl --socket /tmp/receptor.sock work submit --rm --follow --no-payload --param secret_kube_config=@/etc/rancher/k3s/k3s.yaml --param secret_kube_pod=@./pod-multiple-5.yml demo
WORKER DONE
(MIIBZWKO, released)
```

### 🚫 Failed: Multiple containers in the pod, `worker` container invokes `sleep 0`

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: demo
  namespace: receptor
spec:
  containers:
    - name: worker
      image: busybox
      command: ["sh", "-c", "sleep 0; echo WORKER DONE"]
    - name: sidecar
      image: busybox
      command: ["sh", "-c", "sleep 15; echo SIDECAR DONE"]
  restartPolicy: Never
```

```bash
$ receptorctl --socket /tmp/receptor.sock work submit --rm --follow --no-payload --param secret_kube_config=@/etc/rancher/k3s/k3s.yaml --param secret_kube_pod=@./pod-multiple-0.yml demo
ERROR: Remote unit failed: Error creating pod: expected 1 container in pod but there were 2
(nmBQqMZ3, released)
```

## Additional information

The cases that `worker` invokes `sleep 0`, [we got `ErrPodCompleted` by `UntilWithSync()`](https://github.com/ansible/receptor/blob/8f8481ce94b76ed6d90c8936baf071111275960d/pkg/workceptor/kubernetes.go#L239).
In these cases, [Receptor expects there is only one container in the pod](https://github.com/ansible/receptor/blob/8f8481ce94b76ed6d90c8936baf071111275960d/pkg/workceptor/kubernetes.go#L252-L254) and this is why my fourth case had been failed.
If the pod has only `worker` container, this error was returned but [ignored by caller](https://github.com/ansible/receptor/blob/8f8481ce94b76ed6d90c8936baf071111275960d/pkg/workceptor/kubernetes.go#L288-L294), so the third case completed with success.

Actually I don't have such use case, but I'm reporting what I found while experimenting with various patterns.

### PR DESCRIPTION
**Pull request by kurokobo**
Closes #710

Simply, this PR makes Receptor to find `worker` container and invoke the same validation as implemented before in the case for `ErrPodCompleted`.

Passed the same tests as described in #710.

```bash
$ receptorctl --socket /tmp/receptor.sock work submit --rm --follow --no-payload --param secret_kube_config=@/etc/rancher/k3s/k3s.yaml --param secret_kube_pod=@./pod-single-5.yml demo
WORKER DONE
(RlqiVXB3, released)

$ receptorctl --socket /tmp/receptor.sock work submit --rm --follow --no-payload --param secret_kube_config=@/etc/rancher/k3s/k3s.yaml --param secret_kube_pod=@./pod-single-0.yml demo
WORKER DONE
(SQ1ZviRQ, released)

$ receptorctl --socket /tmp/receptor.sock work submit --rm --follow --no-payload --param secret_kube_config=@/etc/rancher/k3s/k3s.yaml --param secret_kube_pod=@./pod-multiple-5.yml demo
WORKER DONE
(x16GudEc, released)

$ receptorctl --socket /tmp/receptor.sock work submit --rm --follow --no-payload --param secret_kube_config=@/etc/rancher/k3s/k3s.yaml --param secret_kube_pod=@./pod-multiple-0.yml demo
WORKER DONE
(e44l72jZ, released)
```

In addition to above, the patterns that container exited with non-zero exit code are also tested;

```bash
# Exit code of sidecar is ignored and work succeeded
$ cat ./pod-multiple-0-sidecar-failed.yml
apiVersion: v1
kind: Pod
metadata:
  name: demo
  namespace: receptor
spec:
  containers:
    - name: worker
      image: busybox
      command: ["sh", "-c", "sleep 0; echo WORKER DONE"]
    - name: sidecar
      image: busybox
      command: ["sh", "-c", "sleep 0; echo SIDECAR DONE; exit 1"]
  restartPolicy: Never

$ receptorctl --socket /tmp/receptor.sock work submit --rm --follow --no-payload --param secret_kube_config=@/etc/rancher/k3s/k3s.yaml --param secret_kube_pod=@./pod-multiple-0-sidecar-failed.yml demo
WORKER DONE
(fhGshOSJ, released)

# Exit code of worker validated and work failed
$ cat ./pod-multiple-0-worker-failed.yml
apiVersion: v1
kind: Pod
metadata:
  name: demo
  namespace: receptor
spec:
  containers:
    - name: worker
      image: busybox
      command: ["sh", "-c", "sleep 0; echo WORKER DONE; exit 1"]
    - name: sidecar
      image: busybox
      command: ["sh", "-c", "sleep 0; echo SIDECAR DONE"]
  restartPolicy: Never

$ receptorctl --socket /tmp/receptor.sock work submit --rm --follow --no-payload --param secret_kube_config=@/etc/rancher/k3s/k3s.yaml --param secret_kube_pod=@./pod-multiple-0-worker-failed.yml demo
ERROR: Remote unit failed: Error creating pod: container failed with exit code 1: 

(qucjq6N2, released)
```

```bash
# ErrImagePull
$ cat ./pod-multiple-0-errimagepull.yml
apiVersion: v1
kind: Pod
metadata:
  name: demo
  namespace: receptor
spec:
  containers:
    - name: worker
      image: dummy-nonexisting-image
      command: ["sh", "-c", "sleep 0; echo WORKER DONE"]
    - name: sidecar
      image: busybox
      command: ["sh", "-c", "sleep 0; echo SIDECAR DONE"]
  restartPolicy: Never

$ receptorctl --socket /tmp/receptor.sock work submit --rm --follow --no-payload --param secret_kube_config=@/etc/rancher/k3s/k3s.yaml --param secret_kube_pod=@./pod-multiple-0-errimagepull.yml demo
Warning: receptorctl and receptor are different versions, they may not be compatible
ERROR: Remote unit failed: Error creating pod: container failed to start, ImagePullBackOff

(3yW7l5fr, released)
```

### CODE REVIEW

**Comment by kurokobo** [comment]
In the first place, in the current upstream implementation, exit codes of all containers including `worker` are ignored in non-ErrPodCompleted pattern;

- `sleep 0; echo WORKER DONE; exit 1` for `worker` container _causes_ work failure
  - This is `ErrPodCompleted` pattern that the pod exits _before_ becoming `Running`.
- `sleep 5; echo WORKER DONE; exit 1` for `worker` container _does not cause_ work failure
  - This is non-`ErrPodCompleted` pattern that the pod exits _after_ becoming `Running`

The validation for the exit code for `worker` container remains in this PR in the case for `ErrPodCompleted`. However, it would also be good way to avoid verifying the exit code of the `worker` container in all cases in the first place.

**Comment by kurokobo** [comment]
@shanemcd 
Thanks for moving forward this PR. Updated my branch to fix lint error reported from CI.

File: pkg/workceptor/kubernetes.go [hunk 1]
```diff
@@ -249,21 +249,26 @@ func (kw *kubeUnit) createPod(env map[string]string) error {
 
 	if err == ErrPodCompleted {
 		// Hao: shouldn't we also call kw.Cancel() in these cases?
-		if len(kw.pod.Status.ContainerStatuses) != 1 {
-			return fmt.Errorf("expected 1 container in pod but there were %d", len(kw.pod.Status.ContainerStatuses))
-		}
+		for _, cstat := range kw.pod.Status.ContainerStatuses {
```
Comment (AaronH88) [inline pkg/workceptor/kubernetes.go:255]: If I am reading the corev1.Pod code correctly, both the `Status` and `ContainerStatuses` fields are marked with `omitempty` 
What happens to this functions if `Status = nil` ?

**Comment by AaronH88** [comment]
I like the approach,
Can more tests be added to `pkg/workceptor/kubernetes_test.go` to show this PR working please?

**Comment by kurokobo** [comment]
@AaronH88 
Thanks for your review! And sorry for my delay.
I will look into it a bit to see if I can add tests and investigate about your comments (but I can't get a coherent time frame, so it will take some time).

File: pkg/workceptor/kubernetes.go (hunk 1 above)
Comment (bpedersen2) [inline pkg/workceptor/kubernetes.go:255]: I would love to see this change integrated, as it fixes another problem as well: Running in a cluster with sidecar injection (e.g. istio) enabled. In such a case the  1-container check constantly fails.

**Comment by AaronH88** [comment]
PR #718 added required tests for this, happy to merge.

**Review by AaronH88 — APPROVED** [review]
lgtm

**Comment by kurokobo** [comment]
@AaronH88 @shanemcd 
Huge thanks for working on this! ...And sorry for my lack of activities for tests😢

**Pull request merged**